stew/errorhandling.nim

```diff
@@ -0,0 +1,277 @@
+                            ErrorsTuple, OrigResultType)
+      wrapperBody.add newCall(RaisingType, genForwardingCall(procDef))
+
+    wrapperDef.params[0] = if generateTemplate: ident"untyped"
+                           else: RaisingType
+    wrapperDef.body = wrapperBody
+
+    result = newStmtList(procDef, wrapperDef)
+  else:
+    result = procDef
+
+  storeMacroResult result
+
+macro checkForUnhandledErrors(origHandledErrors, raisedErrors: typed) =
+  # This macro is executed with two tuples:
+  #
+  # 1. The list of errors handled at the call-site which will
+  #    have a line info matching the call-site.
+  # 2. The list of errors that the called function is raising.
+  #    The lineinfo here points to the definition of the function.
+
+  # For accidental reasons, the first tuple will be recognized as a
+  # typedesc, while the second won't be (beware because this can be
+  # considered a bug in Nim):
+  var handledErrors = getTypeInst(origHandledErrors)
+  if handledErrors.kind == nnkBracketExpr:
+    handledErrors = handledErrors[1]
+
+  assert handledErrors.kind == nnkTupleConstr and
+         raisedErrors.kind == nnkTupleConstr
+
+  # Here, we'll store the list of errors that the user missed:
+  var unhandledErrors = newTree(nnkPar)
+
+  # We loop through the raised errors and check whether they have
+  # an appropriate handler:
+  for raised in raisedErrors:
+    block findHandler:
+      template tryFindingHandler(raisedType) =
+        for handled in handledErrors:
+          if sameType(raisedType, handled):
+            break findHandler
+
+      tryFindingHandler raised
+      # A base type of the raised exception may be handled instead
+      for baseType in raised.baseTypes:
+        tryFindingHandler baseType
+
+      unhandledErrors.add raised
+
+  if unhandledErrors.len > 0:
+    let errMsg = "The following errors are not handled: $1" % [unhandledErrors.repr]
+    error errMsg, origHandledErrors
+
+template raising*[E, R](x: Raising[E, R]): R =
+  ## `raising` is used to mark locations in the code that might
+  ## raise exceptions. It disarms the type-safety checks imposed
+  ## by the `errors` pragma.
+  distinctBase(x)
+
+macro chk*[R, E](x: Raising[R, E], handlers: untyped): untyped =
+  ## The `chk` macro can be used in 2 different ways
+  ##
+  ## 1) Try to get the result of an expression. In case of any
+  ##    errors, substitute the result with a default value:
+  ##
+  ##    ```
+  ##    let x = chk(foo(), defaultValue)
+  ##    ```
+  ##
+  ##    We'll handle this case with a simple rewrite to
+  ##
+  ##    ```
+  ##    let x = try: distinctBase(foo())
+  ##            except CatchableError: defaultValue
+  ##    ```
+  ##
+  ## 2) Try to get the result of an expression while providing exception
+  ##    handlers that must cover all possible recoverable errors.
+  ##
+  ##    ```
+  ##    let x = chk foo():
+  ##                KeyError as err: defaultValue
```

Comment:
so this example has style issues that we probably want to forbid in our code: raising exceptions from different abstraction layers. 

One of the issues we have is that the API's leak implementation details from different layers - this is a design flaw in these API because it means that error handling is no longer local, and you can no longer consider each layer separate and related only to the layer directly above and below it - it breaks the whole idea of abstraction and layering and turns the application into one giant module where all code depends on all code.

Reply:
You can easily remap exceptions. I was even considering short-cut syntax for this, but I haven't convinced myself it's appropriate:

```nim
let x = chk foo():
        OSError: defaultValue
        ValueError -> MyError(msg: "Some Message")
```